Kate/splitter

jihyeonyi opened the pull request:
<!-- Contributing guide: https://github.com/openvinotoolkit/datumaro/blob/develop/CONTRIBUTING.md -->

### Summary
Added task-specific annotation splitters.
- SplitforClassification
- SplitforMatchingReID
- SplitforDetection

### How to test
```bash
python -m unittest -v tests/test_splitter.py
```

### Checklist
<!-- Put an 'x' in all the boxes that apply -->
- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+    def _split_indice(self, indice, group_name, ratio, required):
+        filtered_size = len(indice)
+        if required > filtered_size:
+            log.warning("Not enough samples for filtered group, '{}'".format(
```
Please follow the `%` formatting style, used everywhere else in the project. You can also omit formatting at all, because `logging` can substitute values itself.

jihyeonyi replied:
Thanks for the comment. I'll replace them all.

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+    def __init__(self, dataset, seed):
+        super().__init__(dataset)
+
+        np.random.seed(seed)
```
I think, this should be done closer to the actual use, because a lot can happen after constructor is called.

jihyeonyi replied:
I'll change this too.

jihyeonyi replied:
Random shuffling occurs several times. 
Do you think I really need to set seed whenever I use the shuffling function?

zhiltsov-max replied:
Just make sure the results are reproducible. I suggest moving seed setting as much close to the point of use, as possible.

jihyeonyi replied:
Then, I'll add test codes to check the results are reproducible.

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+
+        id_subsets, id_ratio = self._validate_splits(id_splits)
+
+        groups = set()
```
Let's defer these calculations to the first use of `__iter__`.

jihyeonyi replied:
I don't understand what you mean.. Can you explain more about this?

zhiltsov-max replied:
Currently, there is a lot of computations going on in the construction time, including input dataset traversal etc. I suggest moving these computations to the first use - in `__iter__`. This can be done by saving parameters in constructor and introducing a flag variable like `initialized` or something like this. Then, when splits are actually required, compute them.

Input dataset traversal should be considered heavy operation, so let's try to avoid it, when not necessary.

jihyeonyi replied:
I understand now. Thank you, I'll apply your suggestion.

jihyeonyi replied:
If we apply this lazy initialization, some error checking would be delayed, too. 
It's because some errors can be noticed while iterating over the dataset.
So even if XXXSplit is constructed successfully, it can give an error later. 
I don't know this is good for users. How do you think?

zhiltsov-max replied:
I think we only need to validate input parameters in constructor. If their application to this specific input dataset produces an error - it looks ok for me to know about this when I run the operation.

jihyeonyi replied:
Thank you for your comment. I'll change the code according to your comment.

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+            trainval_ratio.append(val)
+        trainval_ratio = np.array(trainval_ratio)
+        total_ratio = np.sum(trainval_ratio)
+        if total_ratio < 1e-7:
```
This number is used multiple times, maybe it's worth to be extracted as a constant.

jihyeonyi replied:
Thank you. I'll make it to a constant.

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+    - Single label is expected for each DatasetItem.|n
+    - Each label is expected to have attribute representing the person id. |n
+    """
+    _group_map = dict()
```
This will be a global variable at class level.

jihyeonyi replied:
Yes. it is used by 'get_subset_by_group' function, so I make it a global variable.

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+    """
+    Splits dataset into train/val/test set for detection task.|n
+    For detection dataset, each image can have multiple bbox annotations.|n
+    Since one DataItem can't be included in multiple subsets at the same time,
```
Technically, you can copy an item with different subset values, if it is needed. An item is addressed by `(id, subset)`

jihyeonyi replied:
Oh, for the detection task, it should not occur. Because testing should be done with UNSEEN images.

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+
+        ## 1. group by bbox label
+        by_labels = dict()
+        for idx, item in enumerate(self._extractor):
```
Again, lets avoid iterating over input dataset, unless required.

jihyeonyi replied:
It is required. Can you suggest a better way to iterating over the input dataset?

zhiltsov-max commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+                    if not hasattr(ann, 'label') or ann.label is None:
+                        label = None
+                    else:
+                        label = str(ann.label)
```
`annotation.label` is an index in `extractor.categories()[AnnotationType.label]` (of type [LabelCategories](https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/components/extractor.py#L53)). To get the name of a label, do `extractor.categories()[AnnotationType.label][label].name`.

jihyeonyi replied:
I don't need the name itself. ID is enough.
I just want to treat it as a string type.

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,388 @@
+        # config = { "label1": { "attrs": attr1, "counts": counts },
+        #            "label2": { "attrs": attr2, "counts": counts }}
+        def _get_subset():
+            return np.random.choice(['', 'a', 'b'], p = [0.5, 0.3, 0.2])
```
It is better to avoid randomness in tests.

jihyeonyi replied:
I'll consider a different way to do this.

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,388 @@
+                    for _ in range(count):
+                        idx += 1
+                        iterable.append(DatasetItem(
+                            id=str(idx),
```
`str` is applied automatically.

jihyeonyi replied:
ok. I'll remove it.

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,388 @@
+                'label_' + str(label) for label in range(3)
+            )}
+        source = Dataset.from_iterable(iterable, categories)
+        with self.assertRaises(Exception):
```
Consider using at least `assertRaisesRegex` to avoid successful test passing in unexpected failures, like `KeyError` etc. .

jihyeonyi replied:
Thank you for letting me know this. I'll change all the occurence.

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,388 @@
+            annotations.append(Bbox(1,1,2,2, label=kwargs['label_id'],
+                                id=kwargs['ann_id'],
+                                attributes=kwargs['attributes'],
+                                group=kwargs['label_id']))
```
In `COCO` (and in general) `group` is expected to match `id`. A group is supposed to represent a single object, having multiple annotations (polygons, boxes, (key-)points, masks etc.).

jihyeonyi replied:
Oh.. I missed it. You're right. group should be same to ann_id.  Thanks.

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,388 @@
+                for sname in ['train', 'val', 'test']:
+                    subset = actual.get_subset(sname)
+                    subsets[sname] = subset
+                self.assertEqual(train, len(subsets['train']))
```
Consider using `len(get_subset())` directly, as nothing else is needed for checking.

jihyeonyi replied:
Actually.. I need to check other values, but I can't, with the current statistics function.
Anyway, I'll change this, too. Thank you!

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,388 @@
+            AnnotationType.label: LabelCategories.from_iterable(
+                'label_' + str(label) for label in range(3)
+            )}
+        source = Dataset.from_iterable(iterable, categories)
```
You can do the same simpler with:

```python
Dataset.from_iterable([
    DatasetItem(1, annotations=[Label(0), Label(1)]),
    DatasetItem(2, annotations=[Label(0), Label(2)])
], categories=['a', 'b', 'c'])
```

jihyeonyi replied:
Thank you

zhiltsov-max reviewed (changes requested):
Thank you for a great PR!

I'd suggest fixing code formatting (spaces in arguments) and string formatting to keep the project code in a single style.

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,388 @@
+            annotations.append(Bbox(1,1,2,2, attributes=kwargs['attributes']))
+            annotations.append(Label(0, attributes=kwargs['attributes']))
+
+        functions = {
```
I think, this kind of testing is hard to maintain, but let's leave it as is. However, a generic test should also be added.

jihyeonyi replied:
Can you share your thought more on the generic test?

zhiltsov-max replied:
Basically, the same, as these, but with the most complete inputs. For example - full instances from COCO (bbox or polygons, or mask + points, if it matters, in a group) with a couple of attributes and dataset items - depending on the algorithm. You can also look at test coverage, if you need some hints.

jihyeonyi replied:
I see. Do you think I should enter more complicated test sets even though the DetectionSplit (I changed the class name) only cares Bbox annotations?

zhiltsov-max replied:
No, there is no need in that.

jihyeonyi commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,480 @@
+    DEFAULT_SUBSET_NAME)
+
+class TaskSpecificSplitter(Transform):
+    def __init__(self, dataset, seed):
```
I see. 
I'll add the command line feature later for the other classes (SplitterforXXX) because I have not tested it yet.
Can I add the feature later, not in this pull request?

zhiltsov-max replied:
Ok, no problems with this. Actually, they are added automatically, it's just some polishing and customization.

jihyeonyi commented:
I need help regarding complexity. 
I refactor the code and get complexities like below.
```bash
$ python -m mccabe datumaro/plugins/splitter.py --min 10
188:4: 'SplitforMatchingReID.__init__' 16
307:4: 'SplitforMatchingReID._rebalancing' 12
370:4: 'SplitforDetection.__init__' 19
```
But Codacy gives 35 for SplitforMatchingReID.__init__ and 20 for SplitforDetection.__init__
Why they're different? And how can resolve this issue?

zhiltsov-max commented:
Complexity from these tools is not something we look at. Don't bother with this.

jihyeonyi commented on datumaro/plugins/splitter.py:
```diff
@@ -0,0 +1,478 @@
+                                dataset_key=attr_pid)
+
+            # tag using group
+            for idx, item in enumerate(self._extractor):
+                for subset, split in by_groups.items():
+                    if idx in split:
+                        group_id = self._group_map[subset]
+                        item.annotations[0].group = group_id
+                        break
```
Here, I use 'group' property to split 'test' subset into 'query' and 'gallery' set. 
It seems that the default subset doesn't support a hierarchy. 
In the original implementation, we add "Auxiliary" attribute to the annotation. 
Do you think I change the current implementation w.r.t the original one? 
Or do I have to implement the subset hierarchy? 
Please share your opinion.

zhiltsov-max replied:
Yes, there in no exact support for split hierarchy yet. To do what you want, I'd create 2 extra subsets - `test_gallery` and `test_query`, this would look quite natural (at least, in Datumaro). Technically, subsets are extractors themselves, so it is possible to extend the implementation with some nesting. But it would require lots of changes in the importing/exporting code.

jihyeonyi replied:
Actually, the gallery and query sets are not subsets. They're a kind of tag. 
I need more time to resolve this, so let's leave this as it is at the moment.
I'll resolve this issue while applying the command line feature.

zhiltsov-max commented on tests/test_splitter.py:
```diff
@@ -0,0 +1,631 @@
+    def _get_append_bbox(dataset_type):
+        def append_bbox_coco(annotations, **kwargs):
+            annotations.append(
+                Bbox(
```
Please fix the formatting.

jihyeonyi replied:
I've run black and check format with flake8 to follow pep8 format. 
If it is insufficient, can you give me more hints to resolve this?

zhiltsov-max replied:
The code follows the rules in these specs, but it does not mean it is comfortable to read. Splitting a line on each parameter does not help with readability, in my opinion.

jihyeonyi replied:
I see. I've written bbox info in one line like below, but it seems that 'black' modifies these automatically.
```python
Bbox(
    1, 1, 2, 2, 
    label=...
)
```
I should remove 'black' from pre-commit hook or find another way to resolve this.
Anyway, I'll apply your comment. Thanks.

jihyeonyi commented:
I've rebased and squashed some commits. 
I'd like to resolve the gallery/query tagging issue and support command line execution in the future pull request.
